src/app/users/login.ts

```diff
@@ -0,0 +1,131 @@
+import {
+  getFailedLoginAttemptPerIpLimits,
+  getFailedLoginAttemptPerPhoneLimits,
+  yamlConfig,
+} from "@config/app"
+import { RateLimitPrefix } from "@domain/rate-limit"
+import {
+  RateLimiterExceededError,
+  UserLoginIpRateLimiterExceededError,
+  UserLoginPhoneRateLimiterExceededError,
+} from "@domain/rate-limit/errors"
+import { createToken } from "@services/jwt"
+import { UsersRepository } from "@services/mongoose"
+import { PhoneCodesRepository } from "@services/mongoose/phone-code"
+import { getCarrier } from "@services/phone-provider"
+import { RedisRateLimitService } from "@services/rate-limit"
+
+const checkFailedLoginAttemptPerIpLimits = async (
+  ip: IpAddress,
+): Promise<true | RateLimiterExceededError> => {
+  const limiter = RedisRateLimitService({
+    keyPrefix: RateLimitPrefix.failedAttemptLoginIp,
+    limitOptions: getFailedLoginAttemptPerIpLimits(),
+  })
+  const limitOk = await limiter.consume(ip)
+  if (limitOk instanceof RateLimiterExceededError)
+    return new UserLoginIpRateLimiterExceededError()
+  return limitOk
+}
+
+const rewardFailedLoginAttemptPerIpLimits = async (
+  ip: IpAddress,
+): Promise<true | RateLimiterExceededError> => {
+  const limiter = RedisRateLimitService({
+    keyPrefix: RateLimitPrefix.failedAttemptLoginIp,
+    limitOptions: getFailedLoginAttemptPerIpLimits(),
+  })
+  const limitOk = await limiter.reward(ip)
+  if (limitOk instanceof RateLimiterExceededError)
+    return new UserLoginIpRateLimiterExceededError()
+  return limitOk
+}
+
+const checkFailedLoginAttemptPerPhoneLimits = async (
+  phone: PhoneNumber,
+): Promise<true | RateLimiterExceededError> => {
+  const limiter = RedisRateLimitService({
+    keyPrefix: RateLimitPrefix.failedAttemptLoginPhone,
+    limitOptions: getFailedLoginAttemptPerPhoneLimits(),
+  })
+  const limitOk = await limiter.consume(phone)
+  if (limitOk instanceof RateLimiterExceededError)
+    return new UserLoginPhoneRateLimiterExceededError()
+  return limitOk
+}
+
+// TODO is this a candidate for domain
+const isTestAccount = ({ code, phone }: { code: PhoneCode; phone: PhoneNumber }) =>
+  yamlConfig.test_accounts.findIndex((item) => item.phone === phone) !== -1 &&
+  yamlConfig.test_accounts.filter((item) => item.phone === phone)[0].code.toString() ===
+    code.toString()
+
+export const login = async ({
```

Comment:
Can we put `export`ed functions at the top of the file?

Reply:
yes. is it a convention we are using?

Reply:
I always put the publicly exported parts of a file / module / class at the top no matter what language I'm working in. I think the others have also been doing this - at least for the new code. Though we haven't made it an explicit rule I consider it good practice.

Reply:
Taking a note of this as well on my end

Reply:
Most other files seem to have imports after exports and in some cases an `export default <thing>` right at the end as the last line. Not sure what the usual JS convention is here or if there are any gotchas with putting `export` at the top? cc @samerbuna @dolcalmi

Reply:
imports should be at the beginning and after that we can have non-exported/exported functions, I prefer exported functions at the end but maybe is because I have done "some" js projects/libraries. We should define the rule in the next code review.